lib/DiscoveryStreamFeed.jsm

```diff
@@ -203,11 +207,57 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
       type: at.DISCOVERY_STREAM_SPOCS_UPDATE,
       data: {
         lastUpdated: spocs.lastUpdated,
-        spocs: spocs.data,
+        spocs: this.filterSpocs(spocs.data),
       },
     });
   }
 
+  // Filter spocs based on frequency caps
+  filterSpocs(data) {
+    if (data && data.spocs && data.spocs.length) {
+      const {spocs} = data;
+      const impressions = this.readImpressionsPref(SPOC_IMPRESSION_TRACKING_PREF);
+      return {
+        ...data,
+        spocs: spocs.filter(s => this.isBelowFrequencyCap(impressions, s)),
+      };
+    }
+    return data;
+  }
+
+  // Frequency caps are based on campaigns, which may include multiple spocs.
+  // We currently support two types of frequency caps:
+  // - lifetime: Indicates how many times spocs from a campaign can be shown in total
+  // - period: Indicates how many times spocs from a campaign can be shown within a period
+  //
+  // So, for example, the feed configuration below defines that for campaign 1 no more
+  // than 5 spocs can be show in total, and no more than 2 per hour.
+  // "campaign_id": 1,
+  // "caps": {
+  //  "lifetime": 5,
+  //  "campaign": {
+  //    "count": 2,
+  //    "period": 3600
+  //  }
+  // }
+  isBelowFrequencyCap(impressions, spoc) {
+    const campaignImpressions = impressions[spoc.campaign_id];
+    if (!campaignImpressions) {
+      return true;
+    }
+
+    const lifeTimeCap = Math.min(spoc.caps && spoc.caps.lifetime, MAX_LIFETIME_CAP);
```

Comment:
nit: let's add a guard here in case of `spoc.caps.lifetime` is unset, otherwise, `Math.min(undefined, ...)` would return NAN.